src/scanpy/preprocessing/_utils.py

```diff
@@ -33,14 +33,55 @@ def _(X: np.ndarray, *, axis: Literal[0, 1], dtype: DTypeLike) -> np.ndarray:
 def _get_mean_var(
     X: _SupportedArray, *, axis: Literal[0, 1] = 0
 ) -> tuple[NDArray[np.float64], NDArray[np.float64]]:
-    if isinstance(X, sparse.spmatrix):
-        mean, var = sparse_mean_variance_axis(X, axis=axis)
+    if isinstance(X, np.ndarray):
+        n_threads = numba.get_num_threads()
+        mean, var = _compute_mean_var(X, axis=axis, n_threads=n_threads)
     else:
-        mean = axis_mean(X, axis=axis, dtype=np.float64)
-        mean_sq = axis_mean(elem_mul(X, X), axis=axis, dtype=np.float64)
-        var = mean_sq - mean**2
-    # enforce R convention (unbiased estimator) for variance
-    var *= X.shape[axis] / (X.shape[axis] - 1)
+        if isinstance(X, sparse.spmatrix):
+            mean, var = sparse_mean_variance_axis(X, axis=axis)
+        else:
+            mean = axis_mean(X, axis=axis, dtype=np.float64)
+            mean_sq = axis_mean(elem_mul(X, X), axis=axis, dtype=np.float64)
+            var = mean_sq - mean**2
+        # enforce R convention (unbiased estimator) for variance
+        var *= X.shape[axis] / (X.shape[axis] - 1)
+    return mean, var
+
+
+@numba.njit(cache=True, parallel=True)
+def _compute_mean_var(
+    X: _SupportedArray, axis: Literal[0, 1] = 0, n_threads=1
+) -> tuple[NDArray[np.float64], NDArray[np.float64]]:
+    if axis == 0:
+        axis_i = 1
+        sums = np.zeros((n_threads, X.shape[axis_i]), dtype=np.float64)
+        sums_squared = np.zeros((n_threads, X.shape[axis_i]), dtype=np.float64)
+        mean = np.zeros(X.shape[axis_i], dtype=np.float64)
+        var = np.zeros(X.shape[axis_i], dtype=np.float64)
+        n = X.shape[axis]
+        for i in numba.prange(n_threads):
+            for r in range(i, n, n_threads):
+                for c in range(X.shape[axis_i]):
+                    value = X[r, c]
+                    sums[i, c] += value
+                    sums_squared[i, c] += value * value
+        for c in numba.prange(X.shape[axis_i]):
+            sum_ = sums[:, c].sum()
+            mean[c] = sum_ / n
+            var[c] = (sums_squared[:, c].sum() - sum_ * sum_ / n) / (n - 1)
+    else:
+        axis_i = 0
+        mean = np.zeros(X.shape[axis_i], dtype=np.float64)
+        var = np.zeros(X.shape[axis_i], dtype=np.float64)
+        for r in numba.prange(X.shape[0]):
+            for c in range(X.shape[1]):
+                value = X[r, c]
+                mean[r] += value
+                var[r] += value * value
+        for c in numba.prange(X.shape[0]):
+            mean[c] = mean[c] / X.shape[1]
+            var[c] = (var[c] - mean[c] ** 2) / (X.shape[1] - 1)
```

Comment:
Please don’t duplicate identical lines.

```suggestion
    axis_i = 1 - axis
    mean = np.zeros(X.shape[axis_i], dtype=np.float64)
    var = np.zeros(X.shape[axis_i], dtype=np.float64)
    if axis == 0:
        sums = np.zeros((n_threads, X.shape[axis_i]), dtype=np.float64)
        sums_squared = np.zeros((n_threads, X.shape[axis_i]), dtype=np.float64)
        n = X.shape[axis]
        for i in numba.prange(n_threads):
            for r in range(i, n, n_threads):
                for c in range(X.shape[axis_i]):
                    value = X[r, c]
                    sums[i, c] += value
                    sums_squared[i, c] += value * value
        for c in numba.prange(X.shape[axis_i]):
            sum_ = sums[:, c].sum()
            mean[c] = sum_ / n
            var[c] = (sums_squared[:, c].sum() - sum_ * sum_ / n) / (n - 1)
    else:
        for r in numba.prange(X.shape[0]):
            for c in range(X.shape[1]):
                value = X[r, c]
                mean[r] += value
                var[r] += value * value
        for c in numba.prange(X.shape[0]):
            mean[c] = mean[c] / X.shape[1]
            var[c] = (var[c] - mean[c] ** 2) / (X.shape[1] - 1)
```

Reply:
I think we can slim this down a bit. The two different loops need to be separate though